x/xvalidator/validator.go

```diff
@@ -17,7 +23,98 @@ const (
 	envSeparator  = "="
 )
 
-var envNameRegexp = regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9_]*$")
+// Default global value to used from package level.
+var Validate, Translator = New()
+
+// NewWithPrefix returns a new instance of 'validate' with more validation fields prefixed with 'prefix'.
+func NewWithPrefix(prefix string) (*validator.Validate, ut.Translator) {
+	en := en.New()
+	uni := ut.New(en, en)
+	trans, _ := uni.GetTranslator("en")
+	validate := validator.New()
+
+	validate.RegisterValidation("env", IsEnv)
+	validate.RegisterTranslation("env", trans, func(ut ut.Translator) error {
+		return ut.Add("env", "{0} must be a valid env variable name", false)
+	}, func(ut ut.Translator, fe validator.FieldError) string {
+		t, _ := ut.T(fe.Tag(), fe.Field(), prefix)
+		return t
+	})
+
+	validate.RegisterValidation("portmap", IsPortMapping)
+	validate.RegisterTranslation("portmap", trans, func(ut ut.Translator) error {
+		return ut.Add("portmap", "{0} must be a valid port mapping. eg: 80 or 80:80", false)
+	}, func(ut ut.Translator, fe validator.FieldError) string {
+		t, _ := ut.T(fe.Tag(), fe.Field(), prefix)
+		return t
+	})
+
+	validate.RegisterValidation("domain", IsDomainName)
+	validate.RegisterTranslation("domain", trans, func(ut ut.Translator) error {
+		return ut.Add("domain", "{0} must respect domain-style notation. eg: author.name", false)
+	}, func(ut ut.Translator, fe validator.FieldError) string {
+		t, _ := ut.T(fe.Tag(), fe.Field())
+		return t
+	})
+
+	validate.RegisterValidation("hash", IsHash)
+	validate.RegisterTranslation("hash", trans, func(ut ut.Translator) error {
+		return ut.Add("hash", "{0} must be a valid hash in hex format", false)
+	}, func(ut ut.Translator, fe validator.FieldError) string {
+		t, _ := ut.T(fe.Tag(), fe.Field(), prefix)
+		return t
+	})
+
+	validate.RegisterValidation("accaddress", IsAccAddress)
+	validate.RegisterTranslation("accaddress", trans, func(ut ut.Translator) error {
+		return ut.Add("hash", "{0} must be a valid address", false)
+	}, func(ut ut.Translator, fe validator.FieldError) string {
+		t, _ := ut.T(fe.Tag(), fe.Field(), prefix)
+		return t
+	})
+
+	en_translations.RegisterDefaultTranslations(validate, trans)
+	return validate, trans
+}
+
+// New returns a new instance of 'validate' with more validation fields.
+func New() (*validator.Validate, ut.Translator) {
+	return NewWithPrefix("")
+}
+
+// IsHash validates if given field is valid hash.
+// It checks both string and slice of bytes.
+func IsHash(fl validator.FieldLevel) bool {
```

Comment:
`IsHash` should not return false if the hash is not set. if we want to force an hash to be set, then we should use `required` validation.

Reply:
It should, the same was IsDomain works, or any other different validatior. As you said you can use `required` as well as `omitempty` to controls such behaviour.